pkg/storage/allocator.go

```diff
@@ -648,11 +648,16 @@ func filterBehindReplicas(
 		// Raft leader. This is rare enough not to matter.
 		return nil
 	}
-	quorumIndex := getQuorumIndex(raftStatus, 0)
+	// NB: We use raftStatus.Commit instead of getQuorumIndex() because the
+	// latter can return a value that is less than the commit index. This is
+	// useful for Raft log truncation which sometimes wishes to keep those
+	// earlier indexes, but not appropriate for determining which nodes are
+	// behind the actual commit index of the range.
 	candidates := make([]roachpb.ReplicaDescriptor, 0, len(replicas))
 	for _, r := range replicas {
 		if progress, ok := raftStatus.Progress[uint64(r.ReplicaID)]; ok {
-			if progress.Match >= quorumIndex {
+			if progress.State == raft.ProgressStateReplicate &&
+				progress.Match >= raftStatus.Commit {
```

Comment:
`raftStatus.Commit` is the latest index that has been written to the raft log of a quorum of nodes. One subtlety is that the quorum is determined at the time the index was written. For example, consider a range with 1 replica that we right a bunch of log entries to. The commit index can be some value X. If we add another replica, before that replica has been caught up the quorum index will be less than the commit index.